File: backend/api/resume_templates.py
"""
Resume template management endpoints.

This module provides endpoints for managing customizable resume formatting templates
with organization branding, including CRUD operations for creating, reading, updating,
and deleting resume templates with support for ATS-friendly layouts and styling.
"""
import logging
from typing import List, Optional

from fastapi import APIRouter, HTTPException, Query, status
from fastapi.responses import JSONResponse
from pydantic import BaseModel, Field
from sqlalchemy.exc import SQLAlchemyError

from models.resume_template import ResumeTemplate

logger = logging.getLogger(__name__)

router = APIRouter()


class ResumeTemplateCreate(BaseModel):
    """Request model for creating a resume template."""

    organization_id: Optional[str] = Field(None, description="ID of the organization (None for global templates)")
    name: str = Field(..., description="Template name (e.g., 'Modern', 'Classic', 'Creative')", max_length=255)
    description: Optional[str] = Field(None, description="Description of the template style")
    template_type: str = Field(..., description="Type of resume template (e.g., 'modern', 'classic', 'creative', 'ats_friendly')", max_length=100)
    layout_config: Optional[dict] = Field(None, description="Layout configuration (margins, sections, spacing, etc.)")
    style_config: Optional[dict] = Field(None, description="Style configuration (colors, fonts, headings, etc.)")
    section_config: Optional[dict] = Field(None, description="Section configuration (which sections to include and order)")
    preview_url: Optional[str] = Field(None, description="URL to preview image of the template", max_length=512)
    is_default: bool = Field(False, description="Whether this is the default template")
    is_active: bool = Field(True, description="Whether this template is active")
    is_ats_compliant: bool = Field(False, description="Whether this template is ATS-friendly")
    created_by: Optional[str] = Field(None, description="ID of the user creating the template")


class ResumeTemplateUpdate(BaseModel):
    """Request model for updating a resume template."""

    name: Optional[str] = Field(None, description="Template name", max_length=255)
    description: Optional[str] = Field(None, description="Description of the template style")
    layout_config: Optional[dict] = Field(None, description="Layout configuration")
    style_config: Optional[dict] = Field(None, description="Style configuration")
    section_config: Optional[dict] = Field(None, description="Section configuration")
    preview_url: Optional[str] = Field(None, description="URL to preview image", max_length=512)
    is_default: Optional[bool] = Field(None, description="Whether this is the default template")
    is_active: Optional[bool] = Field(None, description="Whether this template is active")
    is_ats_compliant: Optional[bool] = Field(None, description="Whether this template is ATS-friendly")


class ResumeTemplateResponse(BaseModel):
    """Response model for a resume template."""

    id: str = Field(..., description="Unique identifier for the template")
    organization_id: Optional[str] = Field(None, description="ID of the organization (None for global templates)")
    name: str = Field(..., description="Template name")
    description: Optional[str] = Field(None, description="Description of the template style")
    template_type: str = Field(..., description="Type of resume template")
    layout_config: Optional[dict] = Field(None, description="Layout configuration")
    style_config: Optional[dict] = Field(None, description="Style configuration")
    section_config: Optional[dict] = Field(None, description="Section configuration")
    preview_url: Optional[str] = Field(None, description="URL to preview image")
    is_default: bool = Field(..., description="Whether this is the default template")
    is_active: bool = Field(..., description="Whether this template is active")
    is_ats_compliant: bool = Field(..., description="Whether this template is ATS-friendly")
    created_by: Optional[str] = Field(None, description="ID of the user who created the template")
    created_at: str = Field(..., description="Creation timestamp")
    updated_at: str = Field(..., description="Last update timestamp")


class ResumeTemplateListResponse(BaseModel):
    """Response model for listing resume templates."""

    templates: List[ResumeTemplateResponse] = Field(..., description="List of resume templates")
    total_count: int = Field(..., description="Total number of templates")


@router.post(
    "/",
    response_model=ResumeTemplateResponse,
    status_code=status.HTTP_201_CREATED,
    tags=["Resume Templates"],
)
async def create_resume_template(request: ResumeTemplateCreate) -> JSONResponse:
    """
    Create a resume template.

    This endpoint creates a new resume template with support for custom layouts,
    styling, and sections. Templates can be organization-specific or global.

    Args:
        request: Create request with resume template details

    Returns:
        JSON response with created resume template

    Raises:
        HTTPException(422): If validation fails
        HTTPException(500): If database operation fails

    Examples:
        >>> import requests
        >>> data = {
        ...     "name": "Modern Professional",
        ...     "template_type": "modern",
        ...     "description": "Clean modern design with sidebar",
        ...     "is_ats_compliant": True,
        ...     "layout_config": {"margins": "normal", "sections": ["header", "experience", "skills"]},
        ...     "style_config": {"primary_color": "#2563eb", "font": "Arial"}
        ... }
        >>> response = requests.post("http://localhost:8000/api/resume-templates/", json=data)
        >>> response.json()
        {
            "id": "...",
            "name": "Modern Professional",
            "template_type": "modern",
            ...
        }
    """
    try:
        logger.info(f"Creating resume template: {request.name}, type: {request.template_type}")

        # For now, return placeholder response
        # Database integration will be added in a later subtask when we have async session setup
        template_response = {
            "id": "placeholder-id",
            "organization_id": request.organization_id,
            "name": request.name,
            "description": request.description,
            "template_type": request.template_type,
            "layout_config": request.layout_config,
            "style_config": request.style_config,
            "section_config": request.section_config,
            "preview_url": request.preview_url,
            "is_default": request.is_default,
            "is_active": request.is_active,
            "is_ats_compliant": request.is_ats_compliant,
            "created_by": request.created_by,
            "created_at": "2024-01-25T00:00:00Z",
            "updated_at": "2024-01-25T00:00:00Z",
        }

        logger.info(f"Created resume template with ID: {template_response['id']}")

        return JSONResponse(
            status_code=status.HTTP_201_CREATED,
            content=template_response,
        )

    except HTTPException:
        raise
    except Exception as e:
        logger.error(f"Error creating resume template: {e}", exc_info=True)
        raise HTTPException(
            status_code=status.HTTP_500_INTERNAL_SERVER_ERROR,
            detail=f"Failed to create resume template: {str(e)}",
        ) from e


@router.get("/", tags=["Resume Templates"])
async def list_resume_templates(
    organization_id: Optional[str] = Query(None, description="Filter by organization ID (None for global templates)"),
    template_type: Optional[str] = Query(None, description="Filter by template type"),
    is_default: Optional[bool] = Query(None, description="Filter by default status"),
    is_active: Optional[bool] = Query(True, description="Filter by active status"),
    is_ats_compliant: Optional[bool] = Query(None, description="Filter by ATS compliance"),
    limit: int = Query(100, ge=1, le=1000, description="Maximum number of results"),
    offset: int = Query(0, ge=0, description="Number of results to skip"),
) -> JSONResponse:
    """
    List resume templates with optional filters.

    This endpoint retrieves a list of resume templates with support for filtering
    by organization, template type, default status, and ATS compliance.

    Args:
        organization_id: Optional organization ID filter (None for global templates)
        template_type: Optional template type filter
        is_default: Optional default status filter
        is_active: Optional active status filter (default: true)
        is_ats_compliant: Optional ATS compliance filter
        limit: Maximum number of results to return (default: 100, max: 1000)
        offset: Number of results to skip for pagination (default: 0)

    Returns:
        JSON response with list of resume templates

    Raises:
        HTTPException(500): If database operation fails

    Examples:
        >>> import requests
        >>> response = requests.get("http://localhost:8000/api/resume-templates/?is_ats_compliant=true")
        >>> response.json()
        {
            "templates": [...],
            "total_count": 5
        }
    """
    try:
        logger.info(f"Listing resume templates with filters: organization_id={organization_id}, template_type={template_type}, is_ats_compliant={is_ats_compliant}")

        # For now, return placeholder response
        # Database integration will be added in a later subtask when we have async session setup
        templates = []

        response_data = {
            "templates": templates,
            "total_count": len(templates),
        }

        logger.info(f"Retrieved {response_data['total_count']} resume templates")

        return JSONResponse(
            status_code=status.HTTP_200_OK,
            content=response_data,
        )

    except Exception as e:
        logger.error(f"Error listing resume templates: {e}", exc_info=True)
        raise HTTPException(
            status_code=status.HTTP_500_INTERNAL_SERVER_ERROR,
            detail=f"Failed to list resume templates: {str(e)}",
        ) from e


@router.get("/{template_id}", tags=["Resume Templates"])
async def get_resume_template(template_id: str) -> JSONResponse:
    """
    Get a specific resume template by ID.

    This endpoint retrieves a single resume template by its unique identifier,
    including all layout, style, and section configurations.

    Args:
        template_id: Unique identifier of the resume template

    Returns:
        JSON response with resume template details

    Raises:
        HTTPException(404): If template not found
        HTTPException(500): If database operation fails

    Examples:
        >>> import requests
        >>> response = requests.get("http://localhost:8000/api/resume-templates/abc123")
        >>> response.json()
        {
            "id": "abc123",
            "name": "Modern Professional",
            "template_type": "modern",
            ...
        }
    """
    try:
        logger.info(f"Getting resume template: {template_id}")

        # For now, return placeholder response
        # Database integration will be added in a later subtask when we have async session setup
        template_response = {
            "id": template_id,
            "organization_id": None,
            "name": "Modern Professional",
            "description": "Clean modern design with sidebar",
            "template_type": "modern",
            "layout_config": {
                "margins": "normal",
                "sections": ["header", "experience", "education", "skills"]
            },
            "style_config": {
                "primary_color": "#2563eb",
                "font": "Arial",
                "font_size": 11
            },
            "section_config": {
                "header": {"enabled": True, "position": "top"},
                "experience": {"enabled": True, "position": "main"},
                "education": {"enabled": True, "position": "main"},
                "skills": {"enabled": True, "position": "sidebar"}
            },
            "preview_url": "/previews/modern-professional.png",
            "is_default": True,
            "is_active": True,
            "is_ats_compliant": True,
            "created_by": None,
            "created_at": "2024-01-25T00:00:00Z",
            "updated_at": "2024-01-25T00:00:00Z",
        }

        logger.info(f"Retrieved resume template: {template_id}")

        return JSONResponse(
            status_code=status.HTTP_200_OK,
            content=template_response,
        )

    except HTTPException:
        raise
    except Exception as e:
        logger.error(f"Error getting resume template: {e}", exc_info=True)
        raise HTTPException(
            status_code=status.HTTP_500_INTERNAL_SERVER_ERROR,
            detail=f"Failed to get resume template: {str(e)}",
        ) from e


@router.put("/{template_id}", tags=["Resume Templates"])
async def update_resume_template(
    template_id: str,
    request: ResumeTemplateUpdate,
) -> JSONResponse:
    """
    Update a resume template.

    This endpoint updates an existing resume template with new values for
    layout, style, section configurations, or other properties.

    Args:
        template_id: Unique identifier of the resume template
        request: Update request with fields to update

    Returns:
        JSON response with updated resume template

    Raises:
        HTTPException(404): If template not found
        HTTPException(422): If validation fails
        HTTPException(500): If database operation fails

    Examples:
        >>> import requests
        >>> data = {
        ...     "name": "Updated Modern Professional",
        ...     "style_config": {"primary_color": "#dc2626", "font": "Helvetica"}
        ... }
        >>> response = requests.put("http://localhost:8000/api/resume-templates/abc123", json=data)
        >>> response.json()
        {
            "id": "abc123",
            "name": "Updated Modern Professional",
            ...
        }
    """
    try:
        logger.info(f"Updating resume template: {template_id}")

        # For now, return placeholder response
        # Database integration will be added in a later subtask when we have async session setup
        template_response = {
            "id": template_id,
            "organization_id": None,
            "name": request.name or "Modern Professional",
            "description": request.description or "Clean modern design with sidebar",
            "template_type": "modern",
            "layout_config": request.layout_config or {"margins": "normal"},
            "style_config": request.style_config or {"primary_color": "#2563eb"},
            "section_config": request.section_config or {"experience": {"enabled": True}},
            "preview_url": request.preview_url,
            "is_default": request.is_default if request.is_default is not None else True,
            "is_active": request.is_active if request.is_active is not None else True,
            "is_ats_compliant": request.is_ats_compliant if request.is_ats_compliant is not None else True,
            "created_by": None,
            "created_at": "2024-01-25T00:00:00Z",
            "updated_at": "2024-01-25T01:00:00Z",
        }

        logger.info(f"Updated resume template: {template_id}")

        return JSONResponse(
            status_code=status.HTTP_200_OK,
            content=template_response,
        )

    except HTTPException:
        raise
    except Exception as e:
        logger.error(f"Error updating resume template: {e}", exc_info=True)
        raise HTTPException(
            status_code=status.HTTP_500_INTERNAL_SERVER_ERROR,
            detail=f"Failed to update resume template: {str(e)}",
        ) from e


@router.delete("/{template_id}", tags=["Resume Templates"])
async def delete_resume_template(template_id: str) -> JSONResponse:
    """
    Delete a resume template.

    This endpoint permanently deletes a resume template from the system.

    Args:
        template_id: Unique identifier of the resume template

    Returns:
        JSON response confirming deletion

    Raises:
        HTTPException(404): If template not found
        HTTPException(500): If database operation fails

    Examples:
        >>> import requests
        >>> response = requests.delete("http://localhost:8000/api/resume-templates/abc123")
        >>> response.json()
        {
            "message": "Resume template deleted successfully",
            "id": "abc123"
        }
    """
    try:
        logger.info(f"Deleting resume template: {template_id}")

        # For now, return placeholder response
        # Database integration will be added in a later subtask when we have async session setup

        logger.info(f"Deleted resume template: {template_id}")

        return JSONResponse(
            status_code=status.HTTP_200_OK,
            content={
                "message": "Resume template deleted successfully",
                "id": template_id,
            },
        )

    except HTTPException:
        raise
    except Exception as e:
        logger.error(f"Error deleting resume template: {e}", exc_info=True)
        raise HTTPException(
            status_code=status.HTTP_500_INTERNAL_SERVER_ERROR,
            detail=f"Failed to delete resume template: {str(e)}",
        ) from e

File: frontend/src/types/resume-templates.ts
/**
 * Resume Template type definitions
 *
 * This module contains TypeScript interfaces for all resume template
 * related operations including layout, style, and section configurations.
 */

/**
 * Configuration for resume template layout
 */
export interface LayoutConfig {
  /** Page margins (e.g., 'normal', 'wide', 'narrow') */
  margins?: string;
  /** List of sections to include in the template */
  sections?: string[];
  /** Spacing configuration for different elements */
  spacing?: Record<string, number>;
  /** Additional layout properties */
  [key: string]: unknown;
}

/**
 * Configuration for resume template styling
 */
export interface StyleConfig {
  /** Primary color for headings and accents (hex code) */
  primary_color?: string;
  /** Secondary color for supporting elements */
  secondary_color?: string;
  /** Main font family */
  font?: string;
  /** Base font size in points */
  font_size?: number;
  /** Font family for headings */
  heading_font?: string;
  /** Additional style properties */
  [key: string]: unknown;
}

/**
 * Configuration for a single section in the resume template
 */
export interface SectionConfigItem {
  /** Whether this section is enabled */
  enabled?: boolean;
  /** Position of the section (e.g., 'top', 'main', 'sidebar') */
  position?: string;
  /** Order of the section within the layout */
  order?: number;
  /** Additional section properties */
  [key: string]: unknown;
}

/**
 * Configuration for resume template sections
 */
export interface SectionConfig {
  /** Configuration for each named section */
  [sectionName: string]: SectionConfigItem;
}

/**
 * Resume template type enum
 */
export type ResumeTemplateType =
  | 'modern'
  | 'classic'
  | 'creative'
  | 'ats_friendly'
  | 'professional'
  | 'minimal'
  | 'elegant'
  | 'bold';

/**
 * Resume template create request
 */
export interface ResumeTemplateCreate {
  /** ID of the organization (null for global templates) */
  organization_id?: string | null;
  /** Template name (e.g., 'Modern Professional', 'Classic') */
  name: string;
  /** Description of the template style */
  description?: string | null;
  /** Type/category of the template */
  template_type: string;
  /** Layout configuration (margins, sections, spacing, etc.) */
  layout_config?: LayoutConfig | null;
  /** Style configuration (colors, fonts, headings, etc.) */
  style_config?: StyleConfig | null;
  /** Section configuration (which sections to include and order) */
  section_config?: SectionConfig | null;
  /** URL to preview image of the template */
  preview_url?: string | null;
  /** Whether this is the default template */
  is_default?: boolean;
  /** Whether this template is active and available */
  is_active?: boolean;
  /** Whether this template is ATS-friendly */
  is_ats_compliant?: boolean;
  /** ID of the user creating the template */
  created_by?: string | null;
}

/**
 * Resume template update request
 */
export interface ResumeTemplateUpdate {
  /** Template name */
  name?: string;
  /** Description of the template style */
  description?: string | null;
  /** Layout configuration */
  layout_config?: LayoutConfig | null;
  /** Style configuration */
  style_config?: StyleConfig | null;
  /** Section configuration */
  section_config?: SectionConfig | null;
  /** URL to preview image */
  preview_url?: string | null;
  /** Whether this is the default template */
  is_default?: boolean;
  /** Whether this template is active */
  is_active?: boolean;
  /** Whether this template is ATS-friendly */
  is_ats_compliant?: boolean;
}

/**
 * Resume template response
 */
export interface ResumeTemplateResponse {
  /** Unique identifier for the template */
  id: string;
  /** ID of the organization (null for global templates) */
  organization_id: string | null;
  /** Template name */
  name: string;
  /** Description of the template style */
  description: string | null;
  /** Type/category of the template */
  template_type: string;
  /** Layout configuration */
  layout_config: LayoutConfig | null;
  /** Style configuration */
  style_config: StyleConfig | null;
  /** Section configuration */
  section_config: SectionConfig | null;
  /** URL to preview image */
  preview_url: string | null;
  /** Whether this is the default template */
  is_default: boolean;
  /** Whether this template is active */
  is_active: boolean;
  /** Whether this template is ATS-friendly */
  is_ats_compliant: boolean;
  /** ID of the user who created the template */
  created_by: string | null;
  /** Creation timestamp (ISO 8601) */
  created_at: string;
  /** Last update timestamp (ISO 8601) */
  updated_at: string;
}

/**
 * Resume template list response
 */
export interface ResumeTemplateListResponse {
  /** List of resume templates */
  templates: ResumeTemplateResponse[];
  /** Total number of templates */
  total_count: number;
}

/**
 * Query parameters for listing resume templates
 */
export interface ResumeTemplateListParams {
  /** Filter by organization ID (null for global templates) */
  organization_id?: string | null;
  /** Filter by template type */
  template_type?: string | null;
  /** Filter by default status */
  is_default?: boolean;
  /** Filter by active status */
  is_active?: boolean;
  /** Filter by ATS compliance */
  is_ats_compliant?: boolean;
  /** Maximum number of results to return */
  limit?: number;
  /** Number of results to skip for pagination */
  offset?: number;
}

/**
 * Resume template delete response
 */
export interface ResumeTemplateDeleteResponse {
  /** Confirmation message */
  message: string;
  /** ID of the deleted template */
  id: string;
}

/**
 * Template preview request
 */
export interface TemplatePreviewRequest {
  /** Template ID to preview */
  template_id: string;
  /** Optional resume data to populate the preview */
  resume_data?: Record<string, unknown>;
}

/**
 * Template preview response
 */
export interface TemplatePreviewResponse {
  /** URL to the preview image */
  preview_url: string;
  /** Template ID */
  template_id: string;
  /** Generated timestamp */
  generated_at: string;
}

/**
 * Template customization options
 */
export interface TemplateCustomizationOptions {
  /** Primary color override */
  primary_color?: string;
  /** Font family override */
  font?: string;
  /** Font size override */
  font_size?: number;
  /** Enable/disable specific sections */
  sections?: {
    [key: string]: boolean;
  };
  /** Section order customization */
  section_order?: string[];
}
